File: 프로그래머스/그래프/순위.java
//그래프_2번_순위

package 프로그래머스.그래프;

public class 순위 {
    public static void main(String args[]){
        int n = 5;
        int[][] results = {{4,3},{4,2},{3,2},{1,2},{2,5}};
        System.out.println(solution(n, results));
    }

    public static int solution(int n, int[][] results) {
        int answer = 0;
        int[][] graph = new int[n+1][n+1];

        for(int[] result : results){
            int A = result[0];
            int B = result[1];
            graph[A][B] = 1;
            graph[B][A] = -1;
        }

        for(int i=1;i<=n;i++){
            for(int j=1;j<=n;j++){
                for(int k=1;k<=n;k++){
                    if(graph[i][j]==1 && graph[j][k]==1){
                        graph[i][k] = 1;
                        graph[k][i] = -1;
                    }
                    if(graph[i][j]==-1 && graph[j][k]==-1){
                        graph[i][k] = -1;
                        graph[k][i] = 1;
                    }
                }
            }
        }

        for(int i=1;i<=n;i++){
            int count = 0;
            for(int j=1;j<=n;j++){
                if(i==j) count++;
                else if(graph[i][j]!=0) count++;
                else if(graph[i][j]==0) break;
            }
            if(count==n) answer++;
        }
        return answer;
    }
    
}

File: 프로그래머스/그래프/순위.py
def solution(n, results):
    answer = 0
    
    win, lose = [[] for _ in range(n+1)], [[] for _ in range(n+1)]
    for result in results:
        x, y = result[0], result[1]
        win[x].append(y)
        lose[y].append(x)
    
    global rank
    rank = [0]*(n+1)
    for i in range(1, n+1):
        if len(win[i])+len(lose[i])==n-1 and rank[i]==0:
            rank[i] = len(lose[i])+1
            DFS(i, win, lose, n)
    
    for i in range(1, n+1):
        if rank[i]>0:
            answer+=1
    return answer
            
            
def DFS(player, win, lose, n):
    global rank
    for p in win[player]:
        if len(lose[player]) + len(lose[p]) == n-1:
            rank[p] = rank[player]+1
            DFS(p, win, lose, n)
            
    for p in lose[player]:
        if len(win[player]) + len(win[p]) == n-1:
            rank[p] = rank[player]-1
            DFS(p, win, lose, n)
            
    return

print(solution(5, [[4, 3], [4, 2], [3, 2], [1, 2], [2, 5]]))